File: neptune-docs-code-samples/neptunedata-sdk-samples/java/src/main/java/software/amazon/neptune/samples/NeptuneDataClient.java
package software.amazon.neptune.samples;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.json.JsonMapper;
import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import software.amazon.awssdk.core.document.Document;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.neptunedata.NeptunedataClient;
import software.amazon.awssdk.services.neptunedata.model.*;
import software.amazon.awssdk.http.apache.ApacheHttpClient;
import software.amazon.awssdk.core.retry.RetryPolicy;

import java.net.URI;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Neptune Data API Examples for Java
 */
public class NeptuneDataClient {

    /**
     * The AWS Java v2 SDK has specific Pojo configurations that do not work well with
     * a generally defined Jackson ObjectMapper.  The following is to assist with converting
     * the output of some response objects into JSON.
     */
    private static final ObjectMapper objectMapper = JsonMapper.builder()
        .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
        .visibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
        .visibility(PropertyAccessor.GETTER, JsonAutoDetect.Visibility.NONE)
        .build();

    /**
     * Convert AWS SDK Document to Jackson JsonNode
     */
    private static JsonNode documentToJsonNode(Document document) {
        try {
            String jsonString = document.toString();
            return objectMapper.readTree(jsonString);
        } catch (Exception e) {
            throw new RuntimeException("Error converting Document to JsonNode", e);
        }
    }

    /**
     * Main execution method
     */
    public static void main(String[] args) {

        // Configuration - Replace with your actual Neptune cluster details
        String clusterEndpoint = "my-cluster-name.cluster-abcdefgh1234.my-region.neptune.amazonaws.com";
        int port = 8182;
        String region = "my-region";
        
        // Configure Neptune client
        String endpoint = String.format("https://%s:%d", clusterEndpoint, port);
        
        NeptunedataClient client = NeptunedataClient.builder()
                .region(Region.of(region))
                .endpointOverride(URI.create(endpoint))
                .httpClient(ApacheHttpClient.builder()
                        .socketTimeout(Duration.ZERO) // No timeout
                        .connectionTimeout(Duration.ZERO) // No timeout
                        .build())
                .overrideConfiguration(builder -> builder
                        .retryPolicy(RetryPolicy.builder()
                                .numRetries(0) // No retries
                                .build()))
                .build();
        
        try {
            // ========================================
            // Example 1: Execute Gremlin Query
            // ========================================
            LocalDateTime currentTime = LocalDateTime.now();
            String gremlinQuery = String.format("""
                g.mergeV([(T.label):'person']).
                    option(Merge.onCreate,[name:'alice', age:10]).
                    option(Merge.onMatch,[lastUpdated:single(datetime('%s'))])
                """,currentTime.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
            try {
                ExecuteGremlinQueryRequest gremlinRequest = ExecuteGremlinQueryRequest.builder()
                        .gremlinQuery(gremlinQuery)
                        .serializer("application/vnd.gremlin-v2.0+json;types=false")
                        .build();
                
                ExecuteGremlinQueryResponse gremlinResponse = client.executeGremlinQuery(gremlinRequest);
                JsonNode gremlinResult = documentToJsonNode(gremlinResponse.result());
                System.out.println("Gremlin Result: " + gremlinResult.toPrettyString());
            } catch (Exception e) {
                System.out.println("Gremlin query failed: " + e.getMessage());
            }

            // ========================================
            // Example 2: Execute openCypher Query
            // ========================================
            String opencypherQuery = "MATCH (n) RETURN n LIMIT 5";
            try {
                ExecuteOpenCypherQueryRequest opencypherRequest = ExecuteOpenCypherQueryRequest.builder()
                        .openCypherQuery(opencypherQuery)
                        .build();
                
                ExecuteOpenCypherQueryResponse opencypherResponse = client.executeOpenCypherQuery(opencypherRequest);
                JsonNode opencypherResult = documentToJsonNode(opencypherResponse.results());
                System.out.println("openCypher Results:" + opencypherResult.toPrettyString());
            } catch (Exception e) {
                System.out.println("openCypher query failed: " + e.getMessage());
            }

            // ========================================
            // Example 3: Get Property Graph Summary
            // ========================================
            try {
                GetPropertygraphSummaryRequest summaryRequest = GetPropertygraphSummaryRequest.builder()
                        .mode(GraphSummaryType.BASIC)
                        .build();
                
                GetPropertygraphSummaryResponse summaryResponse = client.getPropertygraphSummary(summaryRequest);
                JsonNode summaryResult = objectMapper.valueToTree(summaryResponse.payload().graphSummary());
                System.out.println("Graph Summary: " + summaryResult.toPrettyString());
            } catch (Exception e) {
                System.out.println("Property graph summary failed: " + e.getMessage());
            }

            // ========================================
            // Example 4: Bulk Loader Job
            // ========================================
            // Update with the S3 bucket where your bulk load files are located
            // Ensure the IAM role listed here is attached to your Neptune cluster:
            //    https://docs.aws.amazon.com/neptune/latest/userguide/bulk-load-tutorial-IAM-add-role-cluster.html
            String s3Source = "s3://your-bucket/your-data/";
            String iamRole = "arn:aws:iam::your-account:role/NeptuneLoadFromS3";
            
            // Start loader job
            StartLoaderJobRequest loaderRequest = StartLoaderJobRequest.builder()
                    .source(s3Source)
                    .format("csv")
                    .iamRoleArn(iamRole)
                    .s3BucketRegion(region)
                    .failOnError(false)
                    .parallelism("HIGH") // change to parallelism to "OVERSUBSCRIBE" to get max load throughput
                    .build();
            
            StartLoaderJobResponse loaderResponse = client.startLoaderJob(loaderRequest);

            String loadId = loaderResponse.payload().get("loadId");
            System.out.println("Started loader job: " + loadId);
            
            // Monitor job completion
            while (true) {
                GetLoaderJobStatusRequest statusRequest = GetLoaderJobStatusRequest.builder()
                        .loadId(loadId)
                        .build();
                
                GetLoaderJobStatusResponse statusResponse = client.getLoaderJobStatus(statusRequest);
                JsonNode statusResult = documentToJsonNode(statusResponse.payload());
                String jobStatus = statusResult.get("overallStatus").get("status").asText();
                System.out.println("Job status: " + jobStatus);
                
                if ("LOAD_COMPLETED".equals(jobStatus) || 
                    "LOAD_FAILED".equals(jobStatus) || 
                    "LOAD_CANCELLED".equals(jobStatus)) {
                    break;
                }
                
                Thread.sleep(10000);
            }
            
        } catch (Exception e) {
            System.err.println("Error in main execution: " + e.getMessage());
            e.printStackTrace();
        } finally {
            client.close();
            System.exit(0);
        }
    }
}

File: neptune-docs-code-samples/neptunedata-sdk-samples/dotnet/NeptuneDataClient.cs
using Amazon;
using Amazon.Neptunedata;
using Amazon.Neptunedata.Model;
using System.Text.Json;

namespace NeptuneDataplaneSamples
{

    public class Program
    {
        /*
         * Helper method to convert query response objects from Document objects
         * into a stringifiable JSON object.
         */
        private static object ConvertDocument(Amazon.Runtime.Documents.Document doc)
        {
            if (doc.IsNull()) return null!;
            if (doc.IsBool()) return doc.AsBool();
            if (doc.IsString()) return doc.AsString();
            if (doc.IsInt()) return doc.AsInt();
            if (doc.IsLong()) return doc.AsLong();
            if (doc.IsDouble()) return doc.AsDouble();
            
            if (doc.IsList())
            {
                var list = doc.AsList();
                return list.Select(ConvertDocument).ToList();
            }
            
            if (doc.IsDictionary())
            {
                var dict = doc.AsDictionary();
                return dict.ToDictionary(kvp => kvp.Key, kvp => ConvertDocument(kvp.Value));
            }
            
            return doc.ToString();
        }

        public static async Task Main(string[] args)
        {
            // Configuration - Replace with your actual Neptune cluster details
            const string clusterEndpoint = "my-cluster-name.cluster-abcdefgh1234.my-region.neptune.amazonaws.com";
            const int port = 8182;
            const string region = "my-region";

            // Configure Neptune client
            var endpoint = $"https://{clusterEndpoint}:{port}";
            var config = new AmazonNeptunedataConfig
            {
                ServiceURL = endpoint,
                MaxErrorRetry = 0, // No retries
                Timeout = Timeout.InfiniteTimeSpan // No timeout
            };

            using var client = new AmazonNeptunedataClient(config);

            try
            {
                // ========================================
                // Example 1: Execute Gremlin Query
                // ========================================
                var currentTime = DateTime.Now.ToString("O");
                var gremlinQuery = $"""
                    g.mergeV([(T.label):'person']).
                    option(Merge.onCreate,[name:'alice', age:10]).
                    option(Merge.onMatch,[lastUpdated:single(datetime('{currentTime}'))])
                    """;
                try
                {
                    var gremlinRequest = new ExecuteGremlinQueryRequest
                    {
                        GremlinQuery = gremlinQuery,
                        Serializer = "application/vnd.gremlin-v2.0+json;types=false"
                    };

                    var gremlinResponse = await client.ExecuteGremlinQueryAsync(gremlinRequest);
                    var convertedResult = ConvertDocument(gremlinResponse.Result);
                    var gremlinJsonString = JsonSerializer.Serialize(convertedResult, new JsonSerializerOptions { WriteIndented = true });
                    Console.WriteLine($"Gremlin Result: {gremlinJsonString}");

                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Gremlin query failed: {ex.Message}");
                }

                // ========================================
                // Example 2: Execute openCypher Query
                // ========================================
                const string opencypherQuery = "MATCH (n) RETURN n LIMIT 5";
                try
                {
                    var opencypherRequest = new ExecuteOpenCypherQueryRequest
                    {
                        OpenCypherQuery = opencypherQuery
                    };

                    var opencypherResponse = await client.ExecuteOpenCypherQueryAsync(opencypherRequest);
                    var convertedResult = ConvertDocument(opencypherResponse.Results);
                    var opencypherJsonString = JsonSerializer.Serialize(convertedResult, new JsonSerializerOptions { WriteIndented = true });
                    Console.WriteLine($"openCypher Results: {opencypherJsonString}");
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"openCypher query failed: {ex.Message}");
                }

                // ========================================
                // Example 3: Get Property Graph Summary
                // ========================================
                try
                {
                    var summaryRequest = new GetPropertygraphSummaryRequest
                    {
                        Mode = GraphSummaryType.Basic
                    };

                    var summaryResponse = await client.GetPropertygraphSummaryAsync(summaryRequest);
                    var summaryJsonString = JsonSerializer.Serialize(summaryResponse.Payload, new JsonSerializerOptions { WriteIndented = true });
                    Console.WriteLine($"Graph Summary: {summaryJsonString}");
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Property graph summary failed: {ex.Message}");
                }

                // ========================================
                // Example 4: Bulk Loader Job
                // ========================================
                // Update with the S3 bucket where your bulk load files are located
                // Ensure the IAM role listed here is attached to your Neptune cluster:
                //     https://docs.aws.amazon.com/neptune/latest/userguide/bulk-load-tutorial-IAM-add-role-cluster.html
                const string s3Source = "s3://your-bucket/your-data/";
                const string iamRole = "arn:aws:iam::your-account:role/NeptuneLoadFromS3";

                try
                {
                    // Start loader job
                    var loaderRequest = new StartLoaderJobRequest
                    {
                        Source = s3Source,
                        Format = "csv",
                        IamRoleArn = iamRole,
                        S3BucketRegion = region,
                        FailOnError = false,
                        Parallelism = "HIGH" // change to parallelism to 'OVERSUBSCRIBE' to get max load throughput
                    };
                    
                    var loaderResponse = await client.StartLoaderJobAsync(loaderRequest);
                    var loadId = loaderResponse.Payload["loadId"].ToString();
                    Console.WriteLine($"Started loader job: {loadId}");
                    
                    // Monitor job completion
                    while (true)
                    {
                        var statusRequest = new GetLoaderJobStatusRequest { LoadId = loadId };
                        var statusResponse = await client.GetLoaderJobStatusAsync(statusRequest);
                        var jobStatus = statusResponse.Payload.AsDictionary()["overallStatus"].AsDictionary()["status"];
                        Console.WriteLine($"Job status: {jobStatus}");
                        
                        if (jobStatus == "LOAD_COMPLETED" || jobStatus == "LOAD_FAILED" || jobStatus == "LOAD_CANCELLED")
                        {
                            break;
                        }
                        
                        await Task.Delay(10000);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Bulk loader failed: {ex.Message}");
                }
                Console.WriteLine();;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in main execution: {ex.Message}");
            }
        }
    }
}